fth.1
.\" Copyright (c) 2004-2017 Michael Scholz <mi-scholz@users.sourceforge.net>
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\" @(#)fth.1	1.138 9/25/17
.\"
.Dd 2017/09/25
.Dt FTH 1
.Os
.\"
.\" NAME
.\"
.Sh NAME
.Nm fth
.Nd Forth script and extension language
.\"
.\" SYNOPSIS
.\"
.Sh SYNOPSIS
.Nm
.Op Fl DdQqrv
.Op Fl C Ar so\(hylib\(hypath
.Op Fl Ee Ar pattern
.Op Fl F Ar fs
.Op Fl f Ar init\(hyfile
.Op Fl I Ar fs\(hypath
.Op Fl S Qq Ar lib init
.Op Fl s Ar file
.Op Ar
.Nm
.Oo Fl al Oc Op Fl i Op Ar suffix
.Op Fl n No \(ba Fl p
.Fl e Ar pattern
.Op Ar file No \(ba Ar \(hy
.Nm
.Op Fl V
.\"
.\" DESCRIPTION
.\"
.Sh DESCRIPTION
This is the manual page for the interactive and script processing
Forth interpreter
.Nm .
It is based on
.Em Ficl .
This manual page provides information for running
.Nm
in a
.Ux
command line environment.  See
.Xr libfth 3
for Forth as an extension language.
.\"
.\" Argument list processing
.\"
.Ss Argument list processing
.Bl -tag -width MMM -compact
.\"
.\" -a
.\"
.It Fl a
Split every line to array
.Ev *farray*
before evaluation.  The split string is field separator
.Ev *fs* .
Used with
.Fl n
or
.Fl p
option.
.\"
.\" -C
.\"
.It Fl C Ar path
Prepend
.Ar path
to
.Ev *load\(hylib\(hypath* ,
multiple calls possible.
.\"
.\" -D
.\"
.It Fl D
Show backtrace and exit with return code 1 if an exception was found
evaluating the initialization file, running a script, loading
additional source files, or evaluating strings from the command line.
.\"
.\" -d
.\"
.It Fl d
Set global variable
.Ev *fth\(hydebug*
to #t.
.\"
.\" -Ee
.\"
.It Fl E Ar pattern
.It Fl e Ar pattern
Evaluate
.Ar pattern .
After evaluation, if options
.Fl e
was specified, exit, if option
.Fl E
was specified, start read\(hyeval\(hyprint\(hyloop.  Multiple calls possible.
If
.Fl D
was given and evaluation was not successful, show backtrace and exit
with return code 1, otherwise exit with return code 0.  The last or
only
.Fl e Ar pattern
can be considered between
.Dl lambda: <{}> pattern\ ;
and is in compile state!  If multiple
.Fl e Ar pattern
appear, all but the last are in interpret state.
.\"
.\" -F
.\"
.It Fl F Ar fs
Set global variable
.Ev *fs*
to field separator
.Ar fs .
.\"
.\" -f
.\"
.It Fl f Ar file
Take initialization file
.Ar file
instead of
.Pa ~/.fthrc .
Use
.Fl Q
to prevent loading initialization files.
.\"
.\" -I
.\"
.It Fl I Ar path
Prepend
.Ar path
to
.Ev *load\(hypath* ,
multiple calls possible.
.\"
.\" -i
.\"
.It Fl i Op Ar suffix
The
.Fl e Ar pattern
will be applied to files or stdin.  If
.Ar suffix
is specified, a backup file will be created with that suffix.
.\"
.\" -l
.\"
.It Fl l
Remove line ending \en when used with
.Fl n
or
.Fl p
option.
.\"
.\" -n
.\"
.It Fl n
Suppress output of current
.Ev *line* .
If not in\(hyplace
.Fl i ,
start an implicit loop.
.\"
.\" -p
.\"
.It Fl p
Print current
.Ev *line*
to stdout.  If not in\(hyplace
.Fl i ,
start an implicit loop.
.\"
.\" -Q
.\"
.It Fl Q
Neither load the system wide nor the local initialization file.
.\"
.\" -q
.\"
.It Fl q
Do not print banner and set global variable
.Ev *fth\(hyverbose*
to #f.
.\"
.\" -r
.\"
.It Fl r
Start a conventional Forth\(hylike session with data stack left
untouched after pressing
.Aq Return .
Normally the data stack is printed and cleared but
.Fl r
changes this behavior to something like a
.Em read\(hyeval\(hyloop
instead of a
.Em read\(hyeval\(hyprint\(hyloop .
If
.Fl D
was given and evaluation was not successful, show backtrace and exit
with return code 1, otherwise exit with return code 0.
.\"
.\" -S
.\"
.It Fl S Qq Ar lib init
Load C extension library
.Ar lib
and call the initialization
function
.Ar init .
.Ar lib
and
.Ar init
must be enclosed in quotations
.Dl -S Qq dbm Init_dbm
Multiple calls possible.
.\"
.\" -s
.\"
.It Fl s Ar file
Evaluate
.Ar file
as a Forth script and exit.  It doesn't read the initialization file.
The global Forth variables
.Ev *argc*
and
.Ev *argv*
are set and filled with the count and command line options appearing
after
.Fl s Ar file .
This must be the last option given
to the
.Nm
interpreter.  Options after that are taken as script options.  If
.Fl D
was given and evaluation was not successful, show a backtrace and
exit with return code 1, otherwise exit with return code 0.
.\"
.\" -V
.\"
.It Fl V
Display version info and exit.
.\"
.\" -v
.\"
.It Fl v
Set global variable
.Ev *fth\(hyverbose*
to #t (default).
.El
.\"
.\" Forth variables
.\"
.Ss Forth variables
.Bl -tag -width MMM -compact
.\"
.\" *argc*
.\"
.It Ev *argc*
Number of arguments in
.Ev *argv* .
.\"
.\" *argv*
.\"
.It Ev *argv*
Array of command line arguments.
.\"
.\" *exception-list*
.\"
Return array of all exceptions.
.\"
.\" *farray*
.\"
.It Ev *farray*
Auto\(hysplitted array of strings of current line.
.\"
.\" *features*
.\"
.It Ev *features*
Return array of all features.
.\"
.\" *filename*
.\"
.It Ev *filename*
Current filename.
.\"
.\" *fname*
.\"
.It Ev *fname*
Current in\(hyplace filename.
.\"
.\" *fnr*
.\"
.It Ev *fnr*
Input record number in current file.
.\"
.\" *fs*
.\"
.It Ev *fs*
Input field separator, default one space.
.\"
.\" *fth-debug*
.\"
.It Ev *fth\(hydebug*
If #t, show more diagnostic output.
.\"
.\" *fth-verbose*
.\"
.It Ev *fth\(hyverbose*
If #t, show more output.
.\"
.\" *key*
.\"
.It Ev *key*
Hold current value in a map\(emend\(hymap loop.
.\"
.\" *last-exception*
.\"
.It Ev *last\(hyexception*
Hold last raised exception.
.\"
.\" *line*
.\"
.It Ev *line*
Current in\(hyplace processing line.
.\"
.\" *lineno*
.\"
.It Ev *lineno*
Current line number.
.\"
.\" *load-lib-path*
.\"
.It Ev *load\(hylib\(hypath*
Array of directories where
.Nm
searchs for
.Pa *.so
C\(hylibrary files.
.\"
.\" *load-path*
.\"
.It Ev *load\(hypath*
Array of directories where
.Nm
searchs for
.Pa *.fs
script files.
.\"
.\" *loaded-files*
.\"
.It Ev *loaded\(hyfiles*
Array of already loaded files.
.\"
.\" *nr*
.\"
.It Ev *nr*
Input record number over all files.
.\"
.\" *re*
.\"
.It Ev *re*
Array with last regexp match.
.\"
.\" *re-syntax-options*
.\"
.It Ev *re\(hysyntax\(hyoptions*
Regexp syntax option, default
.Dv REG_EXTENDED.
.\"
.\" *re0* ... *re9*
.\"
.It Ev *re0*
Regexp last entire match.
.It Ev *re1* ... *re9*
Regexp last subexpression matches.
.\"
.\" optarg
.\"
.It Ev optarg
Getopt variable with current option argument string, default #f.
.\"
.\" opterr
.\"
.It Ev opterr
If #t, getopt print error message for wrong or missing arguments;
default #t.
.\"
.\" optind
.\"
.It Ev optind
Getopt variable with index of next element in
.Ev *argv*
array, default 1.
.\"
.\" optopt
.\"
.It Ev optopt
Getopt variable with last known option, default #f.
.El
.\"
.\" Hook variables
.\"
.Ss Hook variables
.Nm
calls five hooks if they are not empty: before and after loading
source files and in interactive mode before starting and after
ending the repl as well as every time before showing the prompt.
.Bl -tag -width MMM -compact
.\"
.\" before-load-hook
.\"
.It Cm before-load-hook No (\ filename -- f\ )
Called before loading 
.Ar filename .
If hook returns #f, 
.Ar filename
won't be loaded.
.Bd -literal -offset indent -compact
before-load-hook lambda: <{ fname -- f }>
	\(dq\e\e loading %s\en\(dq #( fname ) fth-print
	#t
; add-hook!
.Ed
.\"
.\" after-load-hook
.\"
.It Cm after-load-hook No (\ filename --\ )
Called after loading 
.Ar filename
and updating global variable 
.Ev *loaded-files* .
.Bd -literal -offset indent -compact
after-load-hook lambda: <{ fname -- }>
	\(dq\e\e %s loaded\en\(dq #( fname ) fth-print
; add-hook!
.Ed
.\"
.\" before-repl-hook
.\"
.It Cm before-repl-hook No (\ --\ )
Called after initializing the
.Xr tecla 7
command-line editing library but before starting the repl.  A
predefined hook showing some help lines can be replaced by your own
message.
.Bd -literal -offset indent -compact
before-repl-hook reset-hook!
before-repl-hook lambda: <{ -- }>
	.\(dq \e\(dq cr
	.\(dq \e Starting FTH on \(dq date .string .\(dq !\(dq cr
	.\(dq \e\(dq cr
; add-hook!
.Ed
.\"
.\" after-repl-hook
.\"
.It Cm after-repl-hook No (\ history-file --\ )
Called after leaving the repl and writing the history file but
before leaving the program.  Its only argument is the history
filename.  You may manipulate the history data entries.  One history
entry consists of two lines: a time stamp preceded by a Forth comment
backslash and the actual history line.
.Bd -literal -offset indent -compact
after-repl-hook lambda: <{ history -- }>
	\e Remove duplicates from history file.
	history readlines array-reverse! { hary }
	#() \(dq\(dq \(dq\(dq { nhary hline tline }
	hary array-length 0 ?do
		hary i    array-ref to hline
		hary i 1+ array-ref to tline
		nhary hline array-member? unless
			nhary hline array-unshift
			tline array-unshift drop
		then
	2 +loop
	history nhary writelines
; add-hook!
.Ed
.\"
.\" before-prompt-hook 
.\"
.It Cm before-prompt-hook No (\ prompt pos -- new-prompt\ )
Called before printing a new prompt to customize the output of it.
.Ar prompt
is the old prompt and 
.Ar pos
the current history position.  The return value, preferable a string,
is the
.Ar prompt
argument for the next hook procedure if any.
.Bd -literal -offset indent -compact
before-prompt-hook lambda: <{ prompt pos -- new-prompt }>
	\(dqfth (%d) ok \(dq #( pos ) string-format
; add-hook!
.Ed
Or with standout and bold mode, see
.Xr gl_prompt_style 3
and
.Sx *promptstyle*
below:
.Bd -literal -offset indent -compact
#t to *promptstyle*
before-prompt-hook lambda: <{ prompt pos -- new-prompt }>
	\(dq%%Sfth (%d)%%s %%Bok%%b \(dq #( pos ) string-format
; add-hook!
.Ed
.El
.\"
.\" The command-line editor
.\"
.Ss The command-line editor
Command-line editing similar to
.Xr tcsh 1
is handled by the
.Xr tecla 7
library if installed.  Predefined key bindings exist for those
similar to
.Xr vi 1
and
.Xr emacs 1 .
The default editing mode is emacs-mode.  You can switch from
emacs-mode to vi-mode via
.Cm M-^V
and from vi-mode to emacs-mode via
.Cm M-^E .
To start in vi-mode, put a line like
.Dl edit-mode vi
in your
.Pa ~/.teclarc
file.  A history of the last command-lines can be listed with
.Cm ^Xh ,
the last 10 history entries can be listed with
.Cm M-10^Xh .
The Tab-key or
.Cm ^I
initiates word completion.  If the Forth dictionary has more than
one entry starting with characters before the cursor, show all, if
only one definition exists, complete it and add a space after the
completed word.  If the Forth dictionary has no entries starting with
the characters before the cursor, try filename completion.  For
complete key listings and function descriptions, see
.Xr tecla 7 .
.Pp
If the first character of the command-line is an exclamation point
.Cm \&! ,
try history expansion similar to
.Xr csh 1 .
.Pp
.Bl -tag -compact
.It !123
repeat event 123
.It !-123
repeat 123rd last event
.It !!
repeat last event (same as !-1)
.It !?str(?)
repeat last event containing
.Ar str;
the last
.Cm \&?
is optional
.It !str
repeat last event starting with
.Ar str
.El
.Pp
If the first character of the command-line is a caret
.Cm ^ ,
try history substitution similar to
.Xr csh 1 .
.Pp
.Bl -tag -compact
.It ^search^replace(^)
repeat last event containing
.Ar search
and replace this string with
.Ar replace ;
the last
.Cm ^
is optional
.El
.\"
.\" Tecla variables
.\"
.Ss Tecla variables

NEWS
See ChangeLog for more.

* 1.3.8

Add config and build tool scripts for scons.  It's a replacement
for autoconfig tools.

Remove command line option -h.

New experimental targets in src/Makefile:
	fth-llvm and fth-llvm.sh
This requires C compiler clang as well as lli and llc from the llvm
project.  These targets are not required for a working fth interpreter.

New global function fth_xt_call and fth_xt_apply added.

New global function fth_regexp_var_ref added.

Global variable *clm-output-device* removed from
examples/site-lib/clm.fs.


trunk/src/fth.c
/*-
 * Copyright (c) 2005-2017 Michael Scholz <mi-scholz@users.sourceforge.net>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#if !defined(lint)
const char fth_sccsid[] = "%W% %G%";
#endif /* not lint */

#if defined(HAVE_CONFIG_H)
#include "config.h"
#endif

#include "fth.h"
#include "utils.h"
#include <getopt.h>

#define FTH_COPYRIGHT	"(c) 2004-2017 Michael Scholz"

static FTH	 eval_with_error_exit(void *p, int kind);
static void	 repl_in_place(char *in, FTH out, ficlWord *word,
		     bool auto_split_p, bool print_p, bool chomp_p);
static ficlWord *source_to_word(const char *buffer);
static FTH	 string_split(char *str, char *delim);

enum {
	REPL_COMPILE,
	REPL_INTERPRET
};

static ficlWord *
source_to_word(const char *buffer)
{
	ficlWord *word = NULL;
	char *bstr;
	int status;

	bstr = fth_format("lambda: ( ?? -- ?? ) %s ;", buffer);
	status = fth_catch_eval(bstr);
	FTH_FREE(bstr);
	switch (status) {
	case FTH_BYE:
		fth_exit(EXIT_SUCCESS);
		break;
	case FTH_ERROR:
		fth_exit(EXIT_FAILURE);
		break;
	default:
		word = ficlStackPopPointer(FTH_FICL_STACK());
		break;
	}
	return (word);
}

/*
 * Like fth_eval() in misc.c but exit(0) if 'bye' and exit(1) if error.
 */
static FTH
eval_with_error_exit(void *p, int kind)
{
	ficlVm *vm;
	FTH val;
	ficlInteger depth, new_depth, i;
	int status;

	val = FTH_UNDEF;
	if (p == NULL)
		return (val);
	fth_eval_p = true;
	vm = FTH_FICL_VM();
	depth = FTH_STACK_DEPTH(vm);
	switch (kind) {
	case REPL_COMPILE:
		status = fth_catch_exec((ficlWord *)p);
		break;
	case REPL_INTERPRET:
	default:
		status = fth_catch_eval((const char *)p);
		break;
	}
	switch (status) {
	case FTH_BYE:
		fth_exit(EXIT_SUCCESS);
		break;
	case FTH_ERROR:
		if (fth_die_on_signal_p)
			fth_exit(EXIT_FAILURE);
		break;
	default:
		new_depth = FTH_STACK_DEPTH(vm) - depth;
		switch (new_depth) {
		case 0:
			val = FTH_UNDEF;
			break;
		case 1:
			val = fth_pop_ficl_cell(vm);
			break;
		default:
			val = fth_make_array_len(new_depth);
			for (i = 0; i < new_depth; i++)
				fth_array_set(val, i, fth_pop_ficl_cell(vm));
			break;
		}
		break;
	}
	fth_eval_p = false;
	return (val);
}

static FTH
string_split(char *str, char *delim)
{
	char *p, *s, *t;
	FTH result;

	s = t = FTH_STRDUP(str);
	result = fth_make_empty_array();
	while ((p = strsep(&s, delim)) != NULL)
		if (*p != '\0')
			fth_array_push(result, fth_make_string(p));
	FTH_FREE(t);
	return (result);
}

static char fth_scratch[BUFSIZ];

static void
repl_in_place(char *in, FTH out, ficlWord *word,
    bool auto_split_p, bool print_p, bool chomp_p)
{
	char *delim, *buf;
	size_t len;
	ficlInteger line_no;
	FTH line;
	FILE *ifp;

	ifp = stdin;
	if (in != NULL) {
		ifp = fopen(in, "r");
		if (ifp == NULL) {
			FTH_SYSTEM_ERROR_ARG_THROW(fopen, in);
			return;
		}
	}
	gc_push(FTH_FICL_VM()->runningWord);
	delim = fth_string_ref(fth_variable_ref("*fs*"));
	line_no = 0;
	buf = fth_scratch;
	while (fgets(buf, BUFSIZ, ifp) != NULL) {
		if (print_p)
			fth_print(buf);
		if (chomp_p) {
			len = fth_strlen(buf);
			if (buf[len - 1] == '\n')
				buf[len - 1] = '\0';
		}
		if (auto_split_p)
			fth_variable_set("*farray*", string_split(buf, delim));
		fth_variable_set("*line*", fth_make_string(buf));
		fth_variable_set("*fnr*", fth_make_int(line_no++));
		line = eval_with_error_exit(word, REPL_COMPILE);
		fth_variable_set("*nr*",
		    fth_number_add(fth_variable_ref("*nr*"), FTH_ONE));
		if (FTH_NOT_FALSE_P(out))
			fth_array_push(out, line);
		gc_loop_reset();
	}
	gc_pop();
	if (in != NULL)
		fclose(ifp);
}

#define LIBSLEN		48
#define WARN_STR	"#<warning: too much calls for -%c, ignoring \"%s\">\n"
#define FTH_USAGE	"\
usage: fth [-DdQqrv] [-C so-lib-path] [-Ee pattern] [-F fs] [-f init-file]\n\
           [-I fs-path] [-S \"lib init\"] [-s file] [file ...]\n\
       fth [-al] [-i [suffix]] [-n | -p] -e pattern [file | -]\n\
       fth -V\n"

extern char    *optarg;
extern int 	opterr;
extern int 	optind;
extern int 	optopt;

int
main(int argc, char **argv)
{
	bool die, no_init_file, auto_split, debug, in_place_p, ficl_repl;
	bool line_end, implicit_loop, loop_print, script_p, finish_getopt;
	int i, c, exit_value, stay_in_repl, verbose;
	int lp_len, llp_len, bufs_len, libs_len;
	char *field_separator, *init_file, *suffix, *script;
	char *buffers[LIBSLEN], *load_lib_paths[LIBSLEN];
	char *libraries[LIBSLEN], *load_paths[LIBSLEN];
	FTH ret;

	/*
	 * environment variable POSIXLY_CORRECT: if set, disable permutation
	 * optstring starting with `-': RETURN_IN_ORDER optstring starting
	 * with `+': REQUIRE_ORDER (posix)
	 * 
	 * optional arguments: append two colons x:: (see i:: in char *args)
	 */
	char *args = "C:DE:F:I:QS:Vade:f:i::lnpqrs:v";

	/*
	 * Long options are gone with version 1.3.3 but --eval and
	 * --no-init-file remain for backwards compatibility for old fth.m4
	 * files.
	 */
	struct option opts[] = {
		{"eval", required_argument, NULL, 'e'},
		{"no-init-file", no_argument, NULL, 'Q'},
		{0, 0, 0, 0}
	};

	exit_value = EXIT_SUCCESS;
	llp_len = 0;		/* -C path */
	die = false;		/* -D */
	/*
	 * stay_in_repl: -1 not set 0 -e || -s 1 -E || -r
	 */
	stay_in_repl = -1;	/* -Er 1 || -es 0 */
	bufs_len = 0;		/* -Ee pattern */
	field_separator = NULL;	/* -F fs */
	lp_len = 0;		/* -I path */
	no_init_file = false;	/* -Q */
	libs_len = 0;		/* -S path */
	auto_split = false;	/* -a */
	debug = false;		/* -d */
	init_file = NULL;	/* -f file */
	in_place_p = false;	/* -i[suffix] */
	suffix = NULL;		/* -isuffix */
	line_end = false;	/* -l */
	implicit_loop = false;	/* -n || -p */
	loop_print = false;	/* -n 0 || -p 1 */
	ficl_repl = false;	/* -r */
	script_p = false;	/* -s */
	finish_getopt = false;	/* -s */
	script = NULL;		/* -s file */
	/*-
	 * verbose: -1 not set --> true in interactive repl
	 *           0 -q quiet
	 *           1 -v verbose
	 */
	verbose = -1;		/* -v 1 || -q 0 */
	opterr = 1;		/* show getopt's error message */
	optind = 1;		/* initialize getopt */

	while (!finish_getopt &&
	    (c = getopt_long(argc, argv, args, opts, NULL)) != -1) {
		switch (c) {
		case 'C':	/* -C PATH */
			if (llp_len < LIBSLEN)
				load_lib_paths[llp_len++] = optarg;
			else
				fprintf(stderr, WARN_STR, c, optarg);
			break;
		case 'D':	/* -D */
			die = true;
			break;
		case 'E':	/* -E PATTERN */
			stay_in_repl = 1;
			if (bufs_len < LIBSLEN)
				buffers[bufs_len++] = optarg;
			else
				fprintf(stderr, WARN_STR, c, optarg);
			break;
		case 'F':	/* -F SEP */
			field_separator = optarg;
			break;
		case 'I':	/* -I PATH */
			if (lp_len < LIBSLEN)
				load_paths[lp_len++] = optarg;
			else
				fprintf(stderr, WARN_STR, c, optarg);
			break;
		case 'Q':	/* -Q */
			no_init_file = true;
			break;
		case 'S':	/* -S "LIB FUNC" */
			if (libs_len < LIBSLEN)
				libraries[libs_len++] = optarg;
			else
				fprintf(stderr, WARN_STR, c, optarg);
			break;
		case 'V':	/* -V */
			fprintf(stdout, "%s %s\n",
			    FTH_PACKAGE_NAME, fth_version());
			fprintf(stdout, "Copyright %s\n", FTH_COPYRIGHT);
			exit(EXIT_SUCCESS);
			break;
		case 'a':	/* -a */
			auto_split = true;
			break;
		case 'd':	/* -d */
			debug = true;
			break;
		case 'e':	/* -e PATTERN */
			stay_in_repl = 0;
			if (bufs_len < LIBSLEN)
				buffers[bufs_len++] = optarg;
			else
				fprintf(stderr, WARN_STR, c, optarg);
			break;
		case 'f':	/* -f FILE */
			init_file = optarg;
			break;
		case 'i':	/* -i [SUFFIX] */
			in_place_p = true;
			if (optarg)
				suffix = optarg;
			break;
		case 'l':	/* -l */
			line_end = true;
			break;
		case 'n':	/* -n */
			loop_print = false;
			implicit_loop = true;
			break;
		case 'p':	/* -p */
			loop_print = true;
			implicit_loop = true;
			break;
		case 'q':	/* -q */
			verbose = 0;
			break;
		case 'r':	/* -r */
			stay_in_repl = 1;
			ficl_repl = true;
			break;
		case 's':	/* -s FILE */
			stay_in_repl = 0;
			script_p = true;
			script = optarg;
			finish_getopt = true;
			break;
		case 'v':	/* -v */
			verbose = 1;
			break;
		case '?':
		default:
			fprintf(stderr, FTH_USAGE);
			exit(EXIT_FAILURE);
			break;
		}
	}

	/*
	 * Start init forth.
	 */
	forth_init_before_load();
	fth_variable_set("*fth-verbose*", BOOL_TO_FTH(verbose > 0));
	fth_variable_set("*fth-debug*", BOOL_TO_FTH(debug));
	fth_die_on_signal_p = die;
	fth_true_repl_p = !ficl_repl;
	/*
	 * If -I or -C was given, we have to add these paths before finishing
	 * Forth init; calling 'make test' before 'make install' requires
	 * this for finding system scripts in the source tree instead in
	 * ${prefix}/share/fth.
	 */
	if (lp_len > 0)		/* -I PATH */
		for (i = 0; i < lp_len; i++)
			fth_unshift_load_path(load_paths[i]);

	if (llp_len > 0)	/* -C PATH */
		for (i = 0; i < llp_len; i++)
			fth_unshift_load_lib_path(load_lib_paths[i]);
	/*
	 * Finish init forth.
	 */
	forth_init();
	/*
	 * Adjust command line array.
	 */
	fth_set_argv(script_p ? optind - 1 : 0, argc, argv);
	argc -= optind;
	argv += optind;
	/*
	 * Reset getopt for further use in Forth scripts.
	 */
	optind = 1;
	if (field_separator != NULL)	/* -F SEP */
		fth_variable_set("*fs*", fth_make_string(field_separator));
	if (libs_len > 0) {	/* -S "LIB FUNC" */
		char *lib, *name, *fnc;

		for (i = 0; i < libs_len; i++) {
			lib = libraries[i];
			name = strsep(&lib, " \t");
			fnc = strsep(&lib, " \t");
			fth_dl_load(name, fnc);
		}
	}
	/*
	 * Run script and exit.
	 */
	if (script_p) {		/* -s FILE */
		ret = fth_load_file(script);
		if (FTH_STRING_P(ret)) {
			fth_throw(FTH_LOAD_ERROR, "%S", ret);
			fth_exit(EXIT_FAILURE);
		}
		fth_exit(EXIT_SUCCESS);
	}
	/*
	 * In-place or implicit-loop action and exit.
	 */
	if (in_place_p || implicit_loop) {	/* -inp */
		ficlWord *word;
		char *in_file, out_file[MAXPATHLEN];
		FTH out;

		if (bufs_len < 1) {
			fth_errorf("#<%s: in-place require -e PATTERN!>\n",
			    fth_exception_ref(FTH_FORTH_ERROR));
			fth_exit(EXIT_FAILURE);
			/*
			 * To silence ccc-analyzer (uninitialized
			 * 'buffer[i]' after for-loop below).
			 */
			/* NOTREACHED */
			return (EXIT_FAILURE);
		}
		/*
		 * If multiple -e, eval all but last.
		 */
		for (i = 0; i < bufs_len - 1; i++)
			eval_with_error_exit(buffers[i], REPL_INTERPRET);
		/*
		 * Last or only -e: compile and use it for in-place.
		 */
		word = source_to_word(buffers[i]);
		/*
		 * Read from stdin ...
		 */
		if (*argv == NULL) {
			repl_in_place(NULL, FTH_FALSE, word,
			    auto_split, loop_print, line_end);
			fth_exit(EXIT_SUCCESS);
		}
		/*
		 * ... or process all remaining files in order.
		 */
		for (i = 0; argv[i]; i++) {
			in_file = argv[i];
			fth_variable_set("*fname*", fth_make_string(in_file));
			if (in_place_p) {	/* -i [SUFFIX] */
				out = fth_make_empty_array();
				repl_in_place(in_file, out, word,
				    auto_split, loop_print, line_end);
				if (suffix != NULL) {	/* -i SUFFIX */
					fth_strcpy(out_file, sizeof(out_file),
					    in_file);
					fth_strcat(out_file, sizeof(out_file),
					    suffix);
					fth_file_rename(in_file, out_file);
				}
				fth_writelines(in_file, out);
			} else
				repl_in_place(in_file, FTH_FALSE, word,
				    auto_split, loop_print, line_end);
		}
		fth_exit(EXIT_SUCCESS);
	}
	/*
	 * Load remaining args as fth source files.
	 */
	for (i = 0; argv[i]; i++) {
		/* read words from stdin and exit */
		if (strcmp(argv[i], "-") == 0) {
			/*-
			 * % echo "80 .f2c cr" | fth -	==> 26.67
			 * 
			 * % cat foo
			 * 80 .f2c cr
			 * 
			 * % fth - < foo		==> 26.67
			 * 
			 * % fth -
			 * 80 .f2c cr <enter>		==> 26.67
			 * bye <enter>
			 * %
			 */
			char *buf;

			buf = fth_scratch;
			while (fgets(buf, BUFSIZ, stdin) != NULL)
				eval_with_error_exit(buf, REPL_INTERPRET);
			fth_exit(EXIT_SUCCESS);
		} else {
			ret = fth_load_file(argv[i]);
			if (FTH_STRING_P(ret)) {
				exit_value++;
				fth_throw(FTH_LOAD_ERROR, "%S", ret);
			}
		}
	}
	/*
	 * Adjust exit-value; if -D, exit.
	 */
	if (exit_value != EXIT_SUCCESS || fth_hit_error_p) {
		exit_value = EXIT_FAILURE;
		if (die)
			fth_exit(exit_value);
	}
	/*
	 * Eval strings from command line.
	 */
	if (bufs_len > 0) {	/* -Ee PATTERN */
		/*
		 * If multiple -e, eval all but last.
		 */
		for (i = 0; i < bufs_len - 1; i++)
			eval_with_error_exit(buffers[i], REPL_INTERPRET);
		/*
		 * Compile last or only -e.
		 */
		eval_with_error_exit(source_to_word(buffers[i]), REPL_COMPILE);
	}
	/*
	 * Adjust stay_in_repl if not set.
	 */
	if (stay_in_repl == -1)
		stay_in_repl = (argc == 0) ? 1 : 0;
	if (stay_in_repl > 0) {
		/*
		 * If not set, be verbose in repl (and while loading init
		 * files).
		 */
		if (verbose == -1)
			fth_variable_set("*fth-verbose*", FTH_TRUE);
		/*
		 * Print banner if we are still here.
		 */
		if (verbose != 0) {	/* not -q */
			fth_printf("\\ This is %s, %s\n",
			    FTH_PACKAGE_NAME, FTH_COPYRIGHT);
			fth_printf("\\ %s %s\n",
			    FTH_PACKAGE_NAME, fth_version());
		}
		/*
		 * Load init files if not -Q.
		 */
		if (!no_init_file) {	/* -f FILE */
			ret = fth_load_global_init_file();
			if (FTH_STRING_P(ret)) {
				exit_value++;
				fth_throw(FTH_LOAD_ERROR, "%S", ret);
			}
			ret = fth_load_init_file(init_file);
			if (FTH_STRING_P(ret)) {
				exit_value++;
				fth_throw(FTH_LOAD_ERROR, "%S", ret);
			}
			if (fth_hit_error_p)
				exit_value = EXIT_FAILURE;
		}
	}
	if (!stay_in_repl || ((exit_value != EXIT_SUCCESS) && die))
		fth_exit(exit_value);
	/*
	 * Finally, start interactive mode.
	 */
	fth_repl(argc, argv);
	return (EXIT_SUCCESS);
}

/*
 * %M% ends here
 */
